### PR TITLE
Fancy output for pipelines and pnid detect

### PR DESCRIPTION
**Pull request by sanderland**

### CODE REVIEW

File: cognite/experimental/data_classes/utils/rules_output.py [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+        for field in sorted(list(matches[0][f"{source_target}KeyFields"].keys()))
+    ]
+    patterns = {
+        (extractor["entitySet"][:-1], extractor["field"]): extractor["pattern"].strip("^$")
+        for extractor in extractors
+        if extractor["extractorType"] == "regex"
+    }
+    html = ""
+    html += (
+        "<table><tr>"  # Look mum I'm a frontend designer
```
Comment (chingyulin) [inline cognite/experimental/data_classes/utils/rules_output.py:66]: fancy

File: cognite/experimental/data_classes/utils/image.py [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+    try:
+        import warnings
+
+        import numpy as np
+        from bounding_box import bounding_box as bb
+        from pdf2image import convert_from_bytes
+        from PIL import Image
+    except ImportError as e:
+        warnings.warn(
+            f"Module {e.name} missing, 'pip install PIL numpy bounding_box pdf2image' for advanced visualization of results"
```
Comment (chingyulin) [inline cognite/experimental/data_classes/utils/image.py:11]:
```suggestion
            f"Module {e.name} missing, 'pip install Pillow numpy bounding_box pdf2image' for advanced visualization of results"
```

File: cognite/experimental/data_classes/utils/image.py [hunk 3]
```diff
@@ -0,0 +1,35 @@
+def draw_boxes(file_bytes, items):
```
Comment (chingyulin) [inline cognite/experimental/data_classes/utils/image.py:1]: We have `convert` endpoint, I think use it with `display(SVG(svg))` from `Ipython` could be an option?

File: cognite/experimental/data_classes/contextualization.py [hunk 4]
```diff
@@ -175,3 +325,57 @@ def schedule_interval(self):
...
+    def _repr_html_(self):
+        df = self.to_pandas()
+        try:
+            from IPython.display import display
+
+            image = draw_boxes(self.file_contents, self.result["items"])
+            display(image)
+            return df._repr_html_()
+        except Exception:
+            return df._repr_html_()
```
Comment (chingyulin) [inline cognite/experimental/data_classes/contextualization.py:360]: return outside try-except?

**Review by chingyulin — CHANGES_REQUESTED** [review]
Some not so fancy suggestions

File: cognite/experimental/data_classes/utils/image.py (hunk 3 above)
Comment (sanderland) [reply cognite/experimental/data_classes/utils/image.py:1]: i'm not sure I understand.

File: cognite/experimental/data_classes/contextualization.py [hunk 5]
```diff
@@ -1,35 +1,165 @@
...
+
+
+class EntityMatchingMatchRuleList(CogniteResourceList):
+    _RESOURCE = EntityMatchingMatchRule
+    _ASSERT_CLASSES = False
+
+    def _repr_html_(self):
+        try:
+            from ipywidgets import interact  # dont want this as dependency
+        except:
```
Comment (chingyulin) [inline cognite/experimental/data_classes/contextualization.py:55]:
```suggestion
        except ImportError:
```

File: cognite/experimental/data_classes/contextualization.py [hunk 6]
```diff
@@ -1,35 +1,165 @@
...
 )
-from cognite.client.exceptions import ModelFailedException
-from cognite.client.utils._auxiliary import to_camel_case
-from typing_extensions import TypedDict
+
+from cognite.experimental.data_classes.utils.pandas import dataframe_summarize
+from cognite.experimental.data_classes.utils.rules_output import _color_matches, _label_groups
+
+
+class EntityMatchingMatchRule(CogniteResource):
```
Comment (rebwib) [inline cognite/experimental/data_classes/contextualization.py:18]: Should this also be part of this PR, can not see it in the changelog.

Comment (sanderland) [reply cognite/experimental/data_classes/contextualization.py:18]: - More specific job types returned from these endpoints, with helper methods that make processing and visualizing output easier.

File: cognite/experimental/_api/pnid_parsing.py [hunk 7]
```diff
@@ -125,5 +125,20 @@ def convert(
             file_id=file_id,
             file_external_id=file_external_id,
             items=items,
-            grayscale=grayscale,
+            grayscale=grayscale,job_cls=PNIDConvertResults
         )
+
+    def ocr(self, file_id: int,) -> PNIDDetectionPageList:
+        """Retrieve the cached raw OCR result. Only works when detect has already been run.
```
Comment (rebwib) [inline cognite/experimental/_api/pnid_parsing.py:132]: Or if you run the vision endpoint, but we can add that when it has been added to the sdk.

**Review by rebwib — APPROVED** [review]
Very nice, looks like you have some black issues.

**Review by chingyulin — APPROVED** [review]
LGTM

**Pull request merged**